### ISSUE TITLE
"Empty" Level Ups / Please FIX Growths

### ISSUE THREAD

**Issue by SapphireTactician** [eliatlarge/FEMultiPlayer-V2#47]
Issue type: `problem` 

#### Description:

When training units, multiple units, Sain, Eliwood, Eirika, Edward and especially Marisa immediately come to mind, where **despite paying EXP, not a single stat increases (*by a whole number anyways*)!!!**

My understanding of why this occurs is because growths are fixed in a manner where stats increase by fractions of a point if their growth rate isn't 100% in a stat. Sadly, this method results in numerous occasions where nothing increments to the next number and the character gets an "empty" level-up where no single stat increases by even 1 point - which not only is a waste of the user's EXP, but then forces the player to give the EXP to another unit (*assuming they have enough EXP to level someone other than the affected unit*) when they want to the buff the desired unit - defeating the point while crippling them at the same time.

While not as bad, there's quite a few levels where only HP goes up as well.

The FE games had safeties to prevent this by either re-rolling until at least one stat increased (*which we can't do since RNG is our enemy with competitive team building*), or forcing a minimum of 3 stats to go up per level (*as seen in the Tellius series when awarding Bonus EXP - which our system is very reminiscent of - to a unit who reached their primary growths stat caps*) instead of using decimal increments while praying something goes up.

I understand that fixed growth are vital to ensure consistency on a competitive level, so I'm not sure what kind of fix to propose with our current system without the need for potentially major rebalancing. Personally I like the Tellius series Bonus EXP system where a minimum of 3 stats increase every level in order to make every level-up meaningful, especially in light of the fact that the training aspect is essentially just that: awarding "Bonus EXP."

Can someone PLEASE do something about this?

-------

**TL;DR:** Decimal based growths sometimes result in Level Ups where not even one stat increases to the next whole number. This wastes the player's EXP, cripples units, and affects all units going from Lv 1 to 2 thus wasting a minimum of 4800 EXP collectively. It also tends to happen a lot from 11 to 12. Marisa experiences this misfortune 3 times before reaching Lv 10. RNG growths aren't an option so the re-rolling that FE does to prevent this is out. Our EXP training system very much resembles the Tellius series Bonus EXP System. Perhaps we should mirror that further and just increase stats by whole numbers instead, with a minimum of 3 stats increasing per level.

-------

#### Examples:

* All units (*or close to it*) when leveling from Lv 1 to Lv 2 - effectively wasting a combined total of 4800 EXP on an 8 unit team.
* Dart, Edward, Eirika, Eliwood, Sain, etc. when Leveling Up from Lv 1 to 2, and 11 to 12.
* Marisa when Leveling Up from Lv 1 to 2, 5 to 6, AND 9 to 10
* The list goes on and on

-------

P.S. Sorry if anyone gets notification spam from all my edits, I'm pretty scatterbrained. Also, I seemingly fail at proofreading today :-/

**Comment by rayrobdod** [comment]
I'll echo Duffle's remark from the reddit thread

> Some characters take less investment than others, and it's your job to find out who.

This is not a problem

…

Fine. My understanding of PoR's fixed growths mode is that each characters base is not a whole number, but some arbitrary fraction between the base and the next integer. That prevents the first level from being empty. But it will not prevent all levels from being non-empty.

**Comment by eliatlarge** [comment]
This is just the way that the game is balanced. Some characters just need empty growths to balance them out with other characters, it's really not an issue. I can see in the main games, you don't want players breaking their gameboys over no stats being raised, so the 3 points/1 random is really just a placebo. There is some discussion of lowering the level cap to 10, which would likely solve this, but until we get serious about it I'm closing this for lack of any real change. I can't tell you how much I appreciate your input -- its far more than many others have done, and brought to light many issues that we would like to see addressed.

**Comment by SapphireTactician** [comment]
I beg you to reconsider this "Won't Fix" decision, at the very least make it so that Lv 2 for all units isn't empty, as this serves as a black for 4800 EXP with no visible gain to an entire 8 unit army, they might as well start at Lv 2 because Level 2 is literally trivial due to this bug.

As rayrobdod suggested, perhaps there should be some base decimal value present so that Lv 2 isn't empty for everyone.

>Fine. My understanding of PoR's fixed growths mode is that each characters base is not a whole number, but some arbitrary fraction between the base and the next integer. That prevents the first level from being empty. But it will not prevent all levels from being empty.

This is why I'm not suggesting PoR's Fixed Growth Mode, but rather Radiant Dawn's BEXP system. 3 stats guaranteed to increase by a whole number per level, development team would obviously decide those 3 should be for each individual level based on their growths, balancing decisions, and average end-game stats. I don't mean to burden the developers, but it's more effective to write what each level up gives on a level by level per unit basis than just letting a flawed addition system lazily do the heavy lifting.

Please reconsider, or at least keep the bug open for further discussion.

**Comment by eliatlarge** [comment]
Maybe whole numbers would be easier to work with. I think we use the original base/growth system because those are the stats we got from the games. If this becomes a thing though, rebalancing likely won't be necessary because we can just scale growths to whatever they get at lvl 20.

**Comment by SapphireTactician** [comment]
As long as spending valuable EXP doesn't ever yield an empty level, I'm happy. Otherwise, the player is throwing EXP out the window for no reason. Again, the worst is at level 2, which affects every unit.

**Comment by eliatlarge** [comment]
maybe if we allowed rounding rather than filling each level?

**Comment by SapphireTactician** [comment]
We can give it a shot, so long as no levels are empty, but wouldn't that be the same is whole number increases?

**Comment by wellme** [comment]
I've experimented a little bit with a BEXP-like implementation of level ups in [wellme:alternative-level-ups](https://github.com/wellme/FEMultiPlayer-V2/tree/alternative-level-ups). From what I've seen, it gets pretty much the same results as the regular growths, except a unit always gets around the same amount of stats per level (On average: the sum of the unit's growths / 100). Units also end up with more stats because the method goes from rounding down to (effectively) rounding to the nearest multiple of 100.

Also, with the current algorithm, Matthew and Guy break the speed cap of 35 by reaching 38 and 36 speed at level 20 respectively (Matthew starts breaking the cap at level 17). This is easily fixable though.
